lib/auth/join_tpm_test.go

```diff
@@ -0,0 +1,344 @@
+				AllowedCAs:   caPool,
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKCertSerial:   "mock-ekcert-serial",
+				EKCertVerified: true,
+			},
+		},
+		{
+			name:          "success, both",
+			setEnterprise: true,
+			assertError:   require.NoError,
+
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				JoinRequest: joinRequest(),
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkCert{
+					EkCert: []byte("mock-ekcert"),
+				},
+				AttestationParams: attParams,
+			},
+			tokenSpec: tokenSpec(nil),
+			wantParams: tpm.ValidateParams{
+				EKCert:       []byte("mock-ekcert"),
+				AttestParams: tpm.AttestationParametersFromProto(attParams),
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKCertSerial:   "mock-ekcert-serial-1",
+				EKPubHash:      "mock-ekpub-hashed-1",
+				EKCertVerified: true,
+			},
+		},
+		{
+			name:          "failure, mismatched ekpub",
+			setEnterprise: true,
+			assertError:   allowRulesNotMatched,
+
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				JoinRequest: joinRequest(),
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkKey{
+					EkKey: []byte("mock-ekpub"),
+				},
+				AttestationParams: attParams,
+			},
+			tokenSpec: tokenSpec(nil),
+			wantParams: tpm.ValidateParams{
+				EKKey:        []byte("mock-ekpub"),
+				AttestParams: tpm.AttestationParametersFromProto(attParams),
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKPubHash: "mock-ekpub-hashed-mismatched!",
+			},
+		},
+		{
+			name:          "failure, mismatched ekcert",
+			setEnterprise: true,
+			assertError:   allowRulesNotMatched,
+
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				JoinRequest: joinRequest(),
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkCert{
+					EkCert: []byte("mock-ekcert"),
+				},
+				AttestationParams: attParams,
+			},
+			tokenSpec: tokenSpec(nil),
+			wantParams: tpm.ValidateParams{
+				EKCert:       []byte("mock-ekcert"),
+				AttestParams: tpm.AttestationParametersFromProto(attParams),
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKCertSerial: "mock-ekcert-serial-mismatched!",
+			},
+		},
+		{
+			name:          "failure, verification",
+			setEnterprise: true,
+			assertError: func(t require.TestingT, err error, i ...interface{}) {
+				assert.ErrorContains(t, err, "capacitor overcharged")
+			},
+
+			initReq: &proto.RegisterUsingTPMMethodInitialRequest{
+				JoinRequest: joinRequest(),
+				Ek: &proto.RegisterUsingTPMMethodInitialRequest_EkCert{
+					EkCert: []byte("mock-ekcert"),
+				},
+				AttestationParams: attParams,
+			},
+			tokenSpec: tokenSpec(nil),
+			wantParams: tpm.ValidateParams{
+				EKCert:       []byte("mock-ekcert"),
+				AttestParams: tpm.AttestationParametersFromProto(attParams),
+			},
+			validateReturnTPM: &tpm.ValidatedTPM{
+				EKCertSerial: "mock-ekcert-serial",
+			},
+			validateReturnErr: trace.AccessDenied("capacitor overcharged"),
```

Comment:
Suggestion: don't capture traces:

```suggestion
			validateReturnErr: errors.New("capacitor overcharged"),
```

Reply:
Not addressed?